observer/probers/dns/dns_conf.go

```diff
@@ -126,6 +141,14 @@ func (c DNSConf) MakeProber() (probers.Prober, error) {
 	}, nil
 }
 
+// Instrument constructs any `prometheus.Collector` objects the `DNSProbe` will
+// need to report its own metrics. A map is returned containing the constructed
+// objects, indexed by the name of the prometheus metric. If no objects were
+// constructed, nil is returned.
+func (c DNSConf) Instrument() map[string]*prometheus.Collector {
```

Comment:
We probably don't need the value to be a pointer here. Once these eventually make it into the Prober they're going to have to be dereferenced when they're asserted as Gauges.

```suggestion
func (c DNSConf) Instrument() map[string]prometheus.Collector {
```

Reply:
As I understand it, we don't ever actually need to dereference the pointers because the methods on the prometheus collectors (for example, [GaugeVec](https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#GaugeVec)) have pointer receivers. Thus probers can simply keep those pointers as-is without ever needing to dereference & copy them.

This saves memory, since only one copy of each collector should ever exist in memory, whereas if these are the actual objects, then each Prober object that's constructed will have its own copy (harmless and would work, but a waste of memory). If a prober really did ever need to dereference its collector(s) for some reason, that could be left up to the implementation of that prober, so that the other probers could still be using this memory-saving strategy.

Reply:
Though if I'm just flat-out missing something, please let me know

Reply:
Hmm nevermind, I immediately realized what you meant with the assertion part after I posted this. I wonder if there's a better way to go about this though. It just feels to me like there doesn't _need_ to be a proper copy of these collectors for each and every prober of that type

Reply:
😄 If you can find a different approach, I’d be keen to see! I took a hard look at how this would need to be implemented before leaving the comment and this was the conclusion that I came to.

Reply:
So I thought up *a way* to accomplish this, but it'd require yet more refactoring, and I'm not sure if it's worth it. Basically we change `Instrument()` to take in a `Configurer` and return a `Configurer`, and its job is to actually assign the necessary collectors to the probe configurer instance created by `UnmarshalSettings`. Then when `MakeProber` is called on the probe configurer instance, it has the collectors it needs (referenced as pointers to the actual types, not the interface `prometheus.Collector`) and it can assign those collectors into the prober with no need for type assertions.

That's all fine, but the question then becomes when do the collectors for a given prober type get constructed and registered? This is perhaps one downside to this approach, but the *construction* would have to be done at the same time the global `Configurer` object for the given prober is constructed, i.e. within the `init()` function. This would mean that there would always be exactly one set of collectors constructed and sitting in memory for each prober type for the lifetime of the program, regardless of whether that prober type ever actually gets used. However, using pointers would mean that the memory needed would be O(1) instead of O(n) where n is the number of probers of the given type

But I didn't yet answer the question of when/how the collectors would be *registered*. In my mind, I suppose we would just add one more new method to the `Configurer` interface called `RegisterMetrics()` or something along those lines, and that method would be called on the global `Configurer` instance in the same circumstances that we currently call `Instrument()`, but passing in the `prometheus.Registerer` object so that the configurer does the registration.

Btw, it should be noted that in this new scheme, `Instrument()` would be called *every* time a prober configurer instance is created, as it would become part of the flow for constructing a useful configurer. So every time `UnmarshalSettings()` is called, the configurer it returns is immediately passed into `Instrument()` and replaced with the configurer returned from that. Alternatively, the job of `Instrument()` could just be left to `UnmarshalSettings()`, since that's sort of the factory method for making prober configurers. Though in that case, I feel like at least a name change for `UnmarshalSettings()` would be warranted so that it doesn't imply that unmarshalling yaml was *all* the method did

Anyways, that's my rather lengthy musing on how we could do this. There are pros and cons to both approaches imo

Reply:
This was one of the ideas that crossed my mind when reading your original PR. I found that preferred the simpler implementation inside `makeMonitors()` rather than doing more work at package import. Ideally, the operations performed at `init()` should remain very basic (simple constructor registration). All of the configuration  parsing, validation, and object construction happening in one loop is more obvious and easier for the next engineer to reason about. We construct a collector because we encountered a prober which utilized it, we `metrics.Register()` that collector because we know we've got at least one prober that needs it, we store it because we know that any subsequent prober of the same type will need a copy of it, and finally we allow those stored collectors to be garbage collectors once we've finished the the construction of every prober that could possibly need a copy.

Reply:
Fine by me! I'll get that refactor pushed then and resolve the remaining comments